### PR TITLE
Add FXIOS-8714 Toolbar Telemetry

### PR DESCRIPTION
**Pull request by thatswinnie**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8714)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/19327)

## :bulb: Description
Adds telemetry to the new toolbar.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by thatswinnie** [comment]
@adudenamedruby 

# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?
_It will help us understand how the toolbar is used._

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
_Establish baselines or measure changes in product or platform quality or performance._

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?
_There are no other methods_

4) Can current instrumentation answer these questions?
_No._

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>toolbar.qr_scan_button_tapped<br>Record event when user has tapped QR scan button </td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.clear_search_buttton_tapped<br>Record event when user has tapped the clear search button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.share_button_tapped<br>Record event when user taps the share button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.refresh_button_tapped<br>Record event when user has tapped refresh button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.reader_mode_button_tapped<br>Record event when user has tapped the reader mode button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.site_info_button_tapped<br>Record event when user has tapped the lock icon to display the tracking protection.</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.back_button_tapped<br>Record event when user has tapped the back button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.forward_button_tapped<br>Record event when user has tapped the forward button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.back_long_press<br>Record event when user has long pressed the back button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.forward_long_press<br>Record event when user has long pressed the forward button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.home_button_tapped<br>Record event when user has tapped the home button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.one_tap_new_tab_button_tapped<br>Record event when user has tapped the new tab button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.one_tap_new_tab_long_press<br>Record event when user has tapped the new tab button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.search_button_tapped<br>Record event when the user has tapped the search button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.tab_tray_button_tapped<br>Record event when the user has tapped the tab tray button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.tab_tray_long_press<br>Record event when the user has long pressed  the tab tray button</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.app_menu_button_tapped<br>Record event when user has tapped the app menu</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
  <tr>
    <td>toolbar.data_clearance_button_tapped<br>Record event when user has tapped the data clearance</td>
    <td>2</td>
    <td>#19327</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
[Firefox iOS Telemetry Dictionary](https://dictionary.telemetry.mozilla.org/apps/firefox_ios)

7) How long will this data be collected?  Choose one of the following:
_Set to expire on 2025-01-01_

8) What populations will you measure?
* Which release channels? _All_
* Which countries? _All_
* Which locales? _All_
* Any other filters?  Please describe in detail below.

9) If this data collection is default on, what is the opt-out mechanism for users?
_The user can opt out_

10) Please provide a general description of how you will analyze this data.
_The data will be analyzed using Glean._

11) Where do you intend to share the results of your analysis?
_The results will be shared internally through Glean and with advertising partners._

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:
_No third-party tools are proposed for this data collection._

File: firefox-ios/Client/metrics.yaml [hunk 1]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
+      is_private:
+        description: Whether the user is in private mode or not
+        type: boolean
+    bugs:
+      - https://github.com/mozilla-mobile/firefox-ios/issues/19327
+    data_reviews:
+      - https://github.com/mozilla-mobile/firefox-ios/pull/22325
+    notification_emails:
+      - fx-ios-data-stewards@mozilla.com
+    expires: "2025-01-01"
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4916]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 2]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4900]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 3]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4884]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 4]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4868]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 5]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4852]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 6]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4836]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 7]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4820]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 8]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4804]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 9]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4788]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 10]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4772]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 11]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4756]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 12]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4740]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 13]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4724]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 14]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
+      enabled:
+        description: Whether reader mode was opened or closed
+        type: boolean
+    bugs:
+      - https://github.com/mozilla-mobile/firefox-ios/issues/19327
+    data_reviews:
+      - https://github.com/mozilla-mobile/firefox-ios/pull/22325
+    notification_emails:
+      - fx-ios-data-stewards@mozilla.com
+    expires: "2025-01-01"
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4708]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 15]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
+      is_private:
+        description: Whether the user is in private mode or not
+        type: boolean
+    bugs:
+      - https://github.com/mozilla-mobile/firefox-ios/issues/19327
+    data_reviews:
+      - https://github.com/mozilla-mobile/firefox-ios/pull/22325
+    notification_emails:
+      - fx-ios-data-stewards@mozilla.com
+    expires: "2025-01-01"
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4689]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 16]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4673]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 17]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4657]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/metrics.yaml [hunk 18]
```diff
@@ -4620,6 +4620,300 @@ awesomebar:
...
```
Comment (adudenamedruby) [inline firefox-ios/Client/metrics.yaml:4641]:
```suggestion
    expires: "2025-06-01"
```

File: firefox-ios/Client/Telemetry/TelemetryWrapper.swift [hunk 19]
```diff
@@ -2115,6 +2140,118 @@ extension TelemetryWrapper {
             }
             GleanMetrics.Pings.shared.fxSuggest.submit()
 
+        // MARK: - Toolbar
```
Comment (adudenamedruby) [inline firefox-ios/Client/Telemetry/TelemetryWrapper.swift:2143]: Uh, you're gonna hate me for this, but, the new method we're looking to implement, telemetry wise, going forward, is creating specific telemetry utilities for anything that uses telemetry in Redux. So when you instantiate your middleware, or your reducer or whatnot, you'll have a telemetry utility for that thing, where you'll do the needful, and thus avoiding embiggening the `TelemetryWrapper`. An example is [MicrosurveyTelemety](https://github.com/mozilla-mobile/firefox-ios/blob/912eae5754061554f7fe94283c7433edf9878b1b/firefox-ios/Client/Frontend/Microsurvey/MicrosurveyTelemetry.swift)

anything that's currently not easy to go through the utility, we can still add in the wrapper, but we'll (hopefully very soon) be getting rid of the convoluted headache that it is.

**Review by adudenamedruby — CHANGES_REQUESTED** [review]

**Comment by adudenamedruby** [comment]
r+
# Data Review Form

_1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?_ 

Yes, telemetry is documented at https://dictionary.telemetry.mozilla.org/apps/firefox_ios

_2) Is there a control mechanism that allows the user to turn the data collection on and off?_ 

Yes, Firefox iOS provides data controls in the app settings

_3) If the request is for permanent data collection, is there someone who will monitor the data over time?_

This will be reviewed by a data-steward on the Firefox iOS team (@adudenamedruby) every 6 months

_4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?_

Type 2

_5) Is the data collection request for default-on or default-off?_

Default on

_6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?_

No

_7) Is the data collection covered by the existing Firefox privacy notice?_ 

Yes

_8) Does the data collection use a third-party collection tool?_ 

No

**Review by adudenamedruby — APPROVED** [review]

**Pull request merged**